File: web-admin/app/meta/validation/assertEvaluator.ts
import type { RuleAssert } from './crossFieldRuleTypes';

type Data = Record<string, unknown>;

export interface AssertResult {
  passed: boolean;
  skipped: boolean;
  failedOperator?: string;
}

const PASSED: AssertResult = { passed: true, skipped: false };
const skipped: AssertResult = { passed: true, skipped: true };
function failed(op: string): AssertResult {
  return { passed: false, skipped: false, failedOperator: op };
}

function resolveValue(value: unknown, data: Data): unknown {
  if (value && typeof value === 'object' && 'ref' in value) {
    return data[(value as { ref: string }).ref];
  }
  return value;
}

function compareOp(left: unknown, right: unknown, op: string): boolean {
  if (right == null) return false;
  if (typeof left === 'number' && typeof right === 'number') {
    switch (op) {
      case 'eq':
        return left === right;
      case 'neq':
        return left !== right;
      case 'gt':
        return left > right;
      case 'gte':
        return left >= right;
      case 'lt':
        return left < right;
      case 'lte':
        return left <= right;
      default:
        return false;
    }
  }
  if (typeof left === 'string' && typeof right === 'string') {
    const cmp = left.localeCompare(right);
    switch (op) {
      case 'eq':
        return cmp === 0;
      case 'neq':
        return cmp !== 0;
      case 'gt':
        return cmp > 0;
      case 'gte':
        return cmp >= 0;
      case 'lt':
        return cmp < 0;
      case 'lte':
        return cmp <= 0;
      default:
        return false;
    }
  }
  return op === 'eq' ? left === right : op === 'neq' ? left !== right : false;
}

/**
 * Evaluate the assert clause of a cross-field rule.
 * Expression mode attempts best-effort JS evaluation.
 */
export function evaluateAssert(assertion: RuleAssert, data: Data): AssertResult {
  if (assertion.expr) {
    try {
      const fn = new Function(...Object.keys(data), `return !!(${assertion.expr})`);
      const result = fn(...Object.values(data));
      return result ? PASSED : failed('expr');
    } catch {
      return skipped; // Expression error → skip
    }
  }

  const field = assertion.field;
  if (!field) return PASSED;

  const value = data[field];

  // Required check — null/empty means FAIL
  if (assertion.required) {
    if (value == null || (typeof value === 'string' && value.trim() === '')) {
      return failed('required');
    }
  }

  // Non-required: null → skip
  if (value == null) return skipped;

  // String constraints
  if (
    assertion.maxLength != null &&
    typeof value === 'string' &&
    value.length > assertion.maxLength
  ) {
    return failed('maxLength');
  }
  if (
    assertion.minLength != null &&
    typeof value === 'string' &&
    value.length < assertion.minLength
  ) {
    return failed('minLength');
  }
  if (
    assertion.pattern != null &&
    typeof value === 'string' &&
    !new RegExp(assertion.pattern).test(value)
  ) {
    return failed('pattern');
  }

  // Comparison operators
  for (const op of ['eq', 'neq', 'gt', 'gte', 'lt', 'lte'] as const) {
    const rhs = assertion[op];
    if (rhs != null) {
      const resolved = resolveValue(rhs, data);
      if (resolved == null) return skipped;
      if (!compareOp(value, resolved, op)) return failed(op);
    }
  }

  // In / NotIn
  if (assertion.in != null) {
    if (!assertion.in.includes(value)) return failed('in');
  }
  if (assertion.notIn != null) {
    if (assertion.notIn.includes(value)) return failed('notIn');
  }

  return PASSED;
}

File: platform/src/main/java/com/auraboot/framework/meta/validation/AssertEvaluator.java
package com.auraboot.framework.meta.validation;

import com.auraboot.framework.meta.dto.RuleAssert;

import java.util.Map;
import java.util.regex.Pattern;

/**
 * Evaluates the 'assert' clause of a cross-field validation rule.
 * Supports declarative mode (field + operators) and expression mode (expr).
 *
 * Null semantics:
 * - required: true + null → FAIL
 * - Other operators + null field → SKIP (treated as passed)
 * - Ref field is null → SKIP
 */
public final class AssertEvaluator {

    private AssertEvaluator() {}

    public static AssertResult evaluate(RuleAssert assertion, Map<String, Object> data) {
        // Expression mode — delegate to SpEL externally
        if (assertion.getExpr() != null) {
            throw new UnsupportedOperationException("Expression asserts must be evaluated externally via SpEL");
        }

        String field = assertion.getField();
        if (field == null) return AssertResult.PASSED;

        Object value = data.get(field);

        // Required check — null/empty means FAIL
        if (Boolean.TRUE.equals(assertion.getRequired())) {
            if (value == null || (value instanceof String s && s.isBlank())) {
                return AssertResult.failed("required");
            }
        }

        // For non-required operators: null field → SKIP
        if (value == null) {
            return AssertResult.SKIPPED;
        }

        // String constraints
        if (assertion.getMaxLength() != null && value instanceof String s) {
            if (s.length() > assertion.getMaxLength()) return AssertResult.failed("maxLength");
        }
        if (assertion.getMinLength() != null && value instanceof String s) {
            if (s.length() < assertion.getMinLength()) return AssertResult.failed("minLength");
        }
        if (assertion.getPattern() != null && value instanceof String s) {
            if (!Pattern.matches(assertion.getPattern(), s)) return AssertResult.failed("pattern");
        }

        // Comparison operators — ref resolution may return null → SKIP
        if (assertion.getEq() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getEq(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "eq")) return AssertResult.failed("eq");
        }
        if (assertion.getNeq() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getNeq(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "neq")) return AssertResult.failed("neq");
        }
        if (assertion.getGt() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getGt(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "gt")) return AssertResult.failed("gt");
        }
        if (assertion.getGte() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getGte(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "gte")) return AssertResult.failed("gte");
        }
        if (assertion.getLt() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getLt(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "lt")) return AssertResult.failed("lt");
        }
        if (assertion.getLte() != null) {
            Object rhs = ConditionEvaluator.resolveValue(assertion.getLte(), data);
            if (rhs == null) return AssertResult.SKIPPED;
            if (!ConditionEvaluator.compareOp(value, rhs, "lte")) return AssertResult.failed("lte");
        }

        // In / NotIn
        if (assertion.getIn() != null) {
            boolean found = assertion.getIn().stream().anyMatch(item -> {
                if (value instanceof Number && item instanceof Number) {
                    return ((Number) value).doubleValue() == ((Number) item).doubleValue();
                }
                return value.equals(item);
            });
            if (!found) return AssertResult.failed("in");
        }
        if (assertion.getNotIn() != null) {
            boolean found = assertion.getNotIn().stream().anyMatch(item -> {
                if (value instanceof Number && item instanceof Number) {
                    return ((Number) value).doubleValue() == ((Number) item).doubleValue();
                }
                return value.equals(item);
            });
            if (found) return AssertResult.failed("notIn");
        }

        return AssertResult.PASSED;
    }

    /**
     * Result of assert evaluation.
     */
    public record AssertResult(boolean passed, boolean skipped, String failedOperator) {
        public static final AssertResult PASSED = new AssertResult(true, false, null);
        public static final AssertResult SKIPPED = new AssertResult(true, true, null);

        public static AssertResult failed(String operator) {
            return new AssertResult(false, false, operator);
        }
    }
}
